src/include/daos/rpc.h

```diff
@@ -105,7 +105,7 @@ enum daos_rpc_type {
 };
 
 /** DAOS_TGT0_OFFSET is target 0's cart context offset */
-#define DAOS_TGT0_OFFSET		(1)
+#define DAOS_TGT0_OFFSET		(2)
```

Comment:
This is one of the things I was wondering about. I still don't have a very deep understanding of the CaRT internals. It seemed like this change was necessary for the definitions of xstream identities to work out, but I'll add Xuezhao to get his input.

Reply:
To save Xuezhao's time, I think we need to gain enough confidence that our approach works, and summarize the properties of our new xstream. For example:

- Should it be a plain pthread or an Argobots xstream? (It needs to run only one "ULT" for the progress loop. Can a plain pthread creates ULTs for the system xstream? If no, then it has to be a xstream.)
- It should be bond to the same CPU with the system xstream, similar to target helper xstreams. (Hence, it should block when there is no dRPC activities.)
- It does not need a CaRT context. (A CaRT context is for fabric communication. Since we listen only to the Unix domain socket, we don't need CaRT contexts.)

Once we work out these, Xuezhao can advise how to add our xstream.

Reply:
DAOS_TGT0_OFFSET is target 0's cart context offset that need used by client, we probably need not change it as drpc is unrelated with cart.

BTW, I did not see where of the change mentioned in the commit msg title "Move dRPC listener to dedicated xstream", as the drpc_listener_run ULT is still running on XS 0 same as before, and the added drpc_handler_ult also runs on XS 0 by this patch.
Do we really need a dedicate XS for drpc? If you want to reserve a dedicate XS for it, maybe can with similar changes like:
#define DRPC_XS_NR (1)
#define dss_sys_xs_nr = (DAOS_TGT0_OFFSET + DRPC_XS_NR)
just FYI.

Reply:
After talking with Johann, I'm leaning toward keeping it an xstream, rather than using a plain pthread. Are there compelling reasons to go outside the Argobots framework for this?

Agreed on the latter two points.

Reply:
> BTW, I did not see where of the change mentioned in the commit msg title "Move dRPC listener to dedicated xstream", as the drpc_listener_run ULT is still running on XS 0 same as before, and the added drpc_handler_ult also runs on XS 0 by this patch.

I updated dss_ult_xs to assign DSS_ULT_DRPC (the listener thread) to XS 1. It seemed that with the number of system XS increased, XS 1 would be our new dedicated ID. Am I missing something further that needs to be done to use the new XS?

One additional question: Is it OK to insert a new XS ID like this and shift the offsets, or should the new ID be added at the end of the list? I.e. Are there any assumptions in code about XS ID?

> Do we really need a dedicate XS for drpc? If you want to reserve a dedicate XS for it, maybe can with similar changes like:
> #define DRPC_XS_NR (1)
> #define dss_sys_xs_nr = (DAOS_TGT0_OFFSET + DRPC_XS_NR)
> just FYI.

I'll try this route. We wanted to add a dedicated XS to reduce latency for dRPC calls, rather than having to yield for some amount of time to avoid hogging XS 0.

Reply:
> I updated dss_ult_xs to assign DSS_ULT_DRPC (the listener thread) to XS 1. It seemed that with the number of system XS increased, XS 1 would be our new dedicated ID. Am I missing something further that needs to be done to use the new XS?
> 
OK, I see the change. only change that is not enough as that will cause the drpc XS reuse the first main IO service XS, you need to create a separate XS for it. You may refer dss_xstreams_init(), dss_start_xs_id() and related code.

> One additional question: Is it OK to insert a new XS ID like this and shift the offsets, or should the new ID be added at the end of the list? I.e. Are there any assumptions in code about XS ID?
> 
Currently the XS0 is for daos internal system service, and followed by a set of IO service XSs. You probably can add XS1 for drpc service XS, and move other IO service XS following/behind the system XS0 and XS1.
May need to pay attention to some details like the XS_ID usage.

Reply:
Sorry, perhaps I'm still missing something. The current dss_xstreams_init() creates system XS in a loop already, based on dss_sys_xs_nr. With dss_sys_xs_nr == 2, the additional system XS should be created with XS_ID 1, right? And the IO service XS should start with XS_ID 2, as the ID calculation relies on the dss_sys_xs_nr value as well. I added some debug output to dss_xstreams_init() that seems to confirm this (system XS 0 and 1, IO service XS 2, offload XS 3 and 4).

Reply:
> After talking with Johann, I'm leaning toward keeping it an xstream, rather than using a plain pthread. Are there compelling reasons to go outside the Argobots framework for this?

No. I was just asking myself if there are compelling reasons to make an xstream. AFAICS, we will not use the Argobots scheduler on this new thread at all. Anyway, I don't have a confident answer; let's go with xstream then.

Reply:
@kjacque, @liuxuezhao, to summarize the properties of the new xstream myself:

- An xstream.
- Bond to the **same** CPU as the system xstream. (Does dss_start_xs_id **really** give us this sharing, or our own CPU?)
- No CaRT context needed.

Reply:
Just a note - I changed the patch to add the dRPC XS to the end of the ID list. I ran daos_test with dRPC using XS ID 1 and found the epoch tests hanging. I'm guessing there's an assumption somewhere in the code about XS ID 1. Some rudimentary searching didn't turn anything up - I'm not totally sure where to look. But changing the dRPC XS ID did resolve the hang.

Reply:
> @kjacque, @liuxuezhao, to summarize the properties of the new xstream myself:
> 
> * An xstream.
agreed

> * Bond to the **same** CPU as the system xstream. (Does dss_start_xs_id **really** give us this sharing, or our own CPU?)
if passing the same xs_id for two dss_start_xs_id() calls then they will bond to same core.
I am not sure if should "Bond to the **same** CPU as the system xstream", if the drpc handling is almost all happen different time (or say non-concurrently) as current system RPC (RDB/iv/bcast etc handling) then seems need not create separate XS for drpc.

> * No CaRT context needed.
agreed.

Reply:
> > * Bond to the **same** CPU as the system xstream. (Does dss_start_xs_id **really** give us this sharing, or our own CPU?)
> 
> if passing the same xs_id for two dss_start_xs_id() calls then they will bond to same core.
> I am not sure if should "Bond to the **same** CPU as the system xstream", if the drpc handling is almost all happen different time (or say non-concurrently) as current system RPC (RDB/iv/bcast etc handling) then seems need not create separate XS for drpc.

@liuxuezhao, it's because we want the dRPC listener to block in poll(2).

Reply:
> I'm guessing there's an assumption somewhere in the code about XS ID 1. Some rudimentary searching didn't turn anything up - I'm not totally sure where to look. But changing the dRPC XS ID did resolve the hang.

I don't know any assumption  about XS ID 1 (as far as I know).
I saw your code that:
"case DSS_ULT_DRPC_HANDLER:
		return 0;
case DSS_ULT_DRPC:
		return DRPC_XS_ID;"

not sure why you handle the drpc in XS 0, but your drpc XS is actually DRPC_XS_ID?

BTW, for your question that "The current dss_xstreams_init() creates system XS in a loop already, based on dss_sys_xs_nr. With dss_sys_xs_nr == 2, the additional system XS should be created with XS_ID 1, right?", I think you are right.

BTW2, I think that if the drpc happens non-concurrently with current system XS, then you need not create separate XS for it. 
If drpc happens concurrently with current system XS tasks, then it is fine to create separate XS for drpc, seems better to keep it in "system XS set" that keep current system XS0, and adding another system XS 1 for drpc. that will make the XS order be more reasonable than adding drpc XS to the end of IO XS. But it is not a super concern for this.

Reply:
> not sure why you handle the drpc in XS 0, but your drpc XS is actually DRPC_XS_ID?

That came out of a previous email discussion. The dRPC listener wants to run on its own XS so poll(2) can block, as Li Wei mentioned. However, some dRPC requests need to call into services like rdb, which are designed to run on XS 0 only. The dRPC handler is what calls the specific method requested by the `Drpc__Call` structure.

Maybe `DSS_ULT_DRPC` should change to `DSS_ULT_DRPC_LISTENER` for clarity.

> seems better to keep it in "system XS set" that keep current system XS0, and adding another system XS 1 for drpc. that will make the XS order be more reasonable than adding drpc XS to the end of IO XS.

That was my intuition as well. I will work on tracking down what's deadlocking the epoch tests when I try to use XS 1.

Reply:
> I will work on tracking down what's deadlocking the epoch tests when I try to use XS 1.

Found this issue. The server was still creating a CaRT context on the dRPC XS, and I think the inconsistency was what was hanging up the test. The tests pass on my system with the latest version.

Reply:
So I think the question remaining in this thread is how to bind the dRPC XS to the same CPU as XS 0.

Reply:
> So I think the question remaining in this thread is how to bind the dRPC XS to the same CPU as XS 0.

@liuxuezhao Can you point me to where we bind XS 0 to its own CPU? I can work from that example.

Reply:
I'd check the hwloc_get_obj_by_depth call in dss_start_xs_id and the hwloc_set_cpubind call in dss_srv_handler.